homeassistant/components/alphaess/sensor.py

```diff
@@ -0,0 +1,160 @@
+        native_unit_of_measurement=ENERGY_KILO_WATT_HOUR,
+        device_class=SensorDeviceClass.ENERGY,
+        state_class=SensorStateClass.TOTAL_INCREASING,
+    ),
+    AlphaESSSensorDescription(
+        key=AlphaESSNames.GridToBattery,
+        name="Grid to Battery",
+        native_unit_of_measurement=ENERGY_KILO_WATT_HOUR,
+        device_class=SensorDeviceClass.ENERGY,
+        state_class=SensorStateClass.TOTAL_INCREASING,
+    ),
+    AlphaESSSensorDescription(
+        key=AlphaESSNames.StateOfCharge,
+        name="State of Charge",
+        native_unit_of_measurement=PERCENTAGE,
+        device_class=SensorDeviceClass.BATTERY,
+        state_class=SensorStateClass.MEASUREMENT,
+    ),
+    AlphaESSSensorDescription(
+        key=AlphaESSNames.Charge,
+        name="Charge",
+        native_unit_of_measurement=ENERGY_KILO_WATT_HOUR,
+        device_class=SensorDeviceClass.ENERGY,
+        state_class=SensorStateClass.TOTAL_INCREASING,
+    ),
+    AlphaESSSensorDescription(
+        key=AlphaESSNames.Discharge,
+        name="Discharge",
+        native_unit_of_measurement=ENERGY_KILO_WATT_HOUR,
+        device_class=SensorDeviceClass.ENERGY,
+        state_class=SensorStateClass.TOTAL_INCREASING,
+    ),
+]
+
+
+async def async_setup_entry(hass, entry, async_add_entities) -> None:
+    """Defer sensor setup to the shared sensor module."""
+
+    coordinator: AlphaESSDataUpdateCoordinator = hass.data[DOMAIN][entry.entry_id]
+
+    entities: List[AlphaESSSensor] = []
+
+    key_supported_states = {
+        description.key: description for description in SENSOR_DESCRIPTIONS
+    }
+
+    for invertor in coordinator.data:
+        serial = invertor
+        for description in key_supported_states:
+            entities.append(
+                AlphaESSSensor(
+                    coordinator, entry, serial, key_supported_states[description].name
+                )
+            )
+    async_add_entities(entities)
+
+    return
+
+
+class AlphaESSSensor(CoordinatorEntity, SensorEntity):
+    """Alpha ESS Base Sensor."""
+
+    def __init__(self, coordinator, config, serial, name):
+        """Initialize the sensor."""
+        super().__init__(coordinator)
+        self._config = config
+        self._name = name
+        self._serial = serial
+        self._coordinator = coordinator
+
+        for invertor in coordinator.data:
+            serial = invertor
+            if self._serial == serial:
+                self._attr_device_info = DeviceInfo(
+                    entry_type=DeviceEntryType.SERVICE,
+                    identifiers={(DOMAIN, serial)},
+                    manufacturer="AlphaESS",
+                    model=coordinator.data[invertor]["Model"],
+                    name=f"Alpha ESS Energy Statistics : {serial}",
+                )
+
+    @property
+    def unique_id(self):
+        """Return a unique ID to use for this entity."""
+        return f"{self._config.entry_id}_{self._serial} - {self._name}"
```

Comment:
I don't think you should use spaces here. Is serial not unique enough? (maybe together with the key)

Reply:
I could use some advice here - AlphaESS supports the idea of an individual owning multiple inverters and solar energy setups.  The serial number is in reference to a specific inverter and then each inverter has multiple sensors which can be queried for various statistics - Solar2Battery, Solar2Load, BatteryState etcetera.  If I just include the serial then how do I differentiate between the various individual sensors.  If I don't include the serial then how do I differentiate between the same sensor for different inverters?  Happy to take best practice for home assistant and try to implement it.

Reply:
If serial is unique per physical device, I would indeed just use `{self.serial}_{self.name}`. 

You can use serial as unique id for a device (not entity), to link all sensors together. Where all entities can use the combination of serial and name.